src/core/profiling/endoscope_backend.c

```diff
@@ -0,0 +1,577 @@
+  for (i = 0; i < GRPC_ENDO_TASK_CAPACITY; i++) {
+    base->task_pool[i].next_task = i + 1;
+    base->task_pool[i].next_taskwithatom = GRPC_ENDO_EMPTY;
+    base->task_pool[i].log_head = GRPC_ENDO_EMPTY;
+  }
+  base->task_pool[GRPC_ENDO_TASK_CAPACITY - 1].next_task = GRPC_ENDO_EMPTY;
+  for (i = 0; i < GRPC_ENDO_THREAD_CAPACITY; i++) {
+    base->thread_pool[i].cycle_created = 0;  /* as new */
+    base->thread_pool[i].timestamp = 0;  /* important */
+  }
+
+  /* error span */
+  base->marker_warning->name = NULL;
+  base->marker_warning->type = 2;  /* EndoMarkerPB::TASK */
+  base->marker_warning->file = "ERROR";
+  base->marker_warning->line = 0;
+  base->marker_warning->function_name = "ERROR";
+  base->marker_warning->timestamp = -1;
+  base->marker_warning->next_marker = GRPC_ENDO_EMPTY;
+
+  base->task_warning->task_id = 0x00ffffff;
+  base->task_warning->marker_id = GRPC_ENDO_MARKER_CAPACITY;
+  base->task_warning->thread_index = GRPC_ENDO_THREAD_CAPACITY;
+  base->task_warning->cycle_end = -1;
+  base->task_warning->log_head = GRPC_ENDO_EMPTY;
+  base->task_warning->log_tail = GRPC_ENDO_EMPTY;
+  base->task_warning->scope_depth = 0;
+  base->task_warning->next_task = GRPC_ENDO_EMPTY;
+  base->task_warning->next_taskwithatom = GRPC_ENDO_EMPTY;
+
+  base->thread_warning->thread_id = 0;
+  base->thread_warning->task_active = GRPC_ENDO_TASK_CAPACITY;
+  base->thread_warning->timestamp = -1;
+}
+
+void grpc_endo_destroy(grpc_endo_base *base) {
+  /* only free dynamic data to prevent leakage */
+#ifdef GRPC_ENDO_COPY_MARKER_NAME
+  GRPC_ENDO_INDEX i;
+  for (i = 0; i < GRPC_ENDO_MARKER_CAPACITY; i++) {
+    free(base->marker_pool[i].name);
+  }
+#endif
+  gpr_mu_destroy(&(base->mutex));
+}
+
+/* Get, Create, and Delete Elements */
+
+static GRPC_ENDO_INDEX grpc_endo_create_marker(grpc_endo_base *base, GRPC_ENDO_INDEX next_index) {
+  /* must be called inside mutex */
+  if (base->marker_count >= GRPC_ENDO_MARKER_CAPACITY) {
+    grpc_endo_warning_print("grpc_endo_create_marker: no marker item available (reached capacity)");
+    grpc_endo_warning_span(base, GRPC_ENDO_WARNING_MARKER);
+    return GRPC_ENDO_EMPTY;
+  } else {
+    GRPC_ENDO_INDEX marker_id = base->marker_count++;
+    /* base->marker_pool[marker_id].name = NULL already initialized */
+    base->marker_pool[marker_id].next_marker = next_index;
+    return marker_id;
+  }
+}
+
+static GRPC_ENDO_INDEX grpc_endo_get_or_create_marker(
+    grpc_endo_base *base, gpr_int32 line, const char *name) {
+  /* thread safe applied */
+  GRPC_ENDO_INDEX *q;
+  gpr_int16 fallback;
+  /* begin hash function */
+  gpr_uint32 hash = line;
+  const char *c;
+  for (c = name; *c != '\0'; c++) {
+    hash += *c;
+    hash += (hash << 10);
+    hash ^= (hash >> 6);
+  }
+  hash += (hash << 3);
+  hash ^= (hash >> 11);
+  hash += (hash << 15);
+  /* end hash function */
+  q = &(base->marker_map[hash % GRPC_ENDO_HASHSIZE]);
+  for (fallback = 100; fallback > 1; fallback--) {
```

Comment:
Also... this needs to be properly thread safe before we can take this pull request (I need to be able to run with TSAN and see no errors).